pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java

```diff
@@ -1272,4 +1389,35 @@ public Timeout getPartitionsAutoUpdateTimeout() {
     }
 
     private static final Logger log = LoggerFactory.getLogger(MultiTopicsConsumerImpl.class);
+
+    public boolean addIgnoreConsumer(ConsumerImpl<T> consumer) {
+        return ignoredConsumers.add(consumer);
+    }
+
+    public boolean removeIgnoreConsumer(ConsumerImpl<T> consumer) {
+        return ignoredConsumers.remove(consumer);
+    }
+
+    public static boolean isIllegalMultiTopicsMessageId(MessageId messageId) {
+        //only support earliest/latest and messageId contains certain partition info
+        if (MessageId.earliest.equals(messageId) || MessageId.latest.equals(messageId)) {
+            return false;
+        }
+        MessageIdImpl messageIdImpl = MessageIdImpl.convertToMessageIdImpl(messageId);
+        if (messageIdImpl != null && messageIdImpl.getPartitionIndex() >= 0 && messageIdImpl.getLedgerId() >= 0
+                && messageIdImpl.getEntryId() >= 0) {
+            return false;
+        }
+        return true;
+    }
+
+    public void tryAcknowledgeMessage(Message<T> msg) {
+        if (msg != null) {
+            BatchMessageIdImpl batchMessageId = new BatchMessageIdImpl(MessageIdImpl.convertToMessageIdImpl(msg.getMessageId()));
+            //Non-batching messages ack every time, batchMessage only need to ack the last one to avoid multi recycle
+            if (batchMessageId.getBatchIndex() < 0 || batchMessageId.getBatchSize() - 1 == batchMessageId.getBatchIndex()) {
+                acknowledgeCumulativeAsync(msg);
```

Comment:
I think here also assume that all topics of the multiple topics consumer are under a partitioned topic?